homeassistant/components/knx/light.py

```diff
@@ -154,19 +236,59 @@ def supported_features(self):
-            flags |= SUPPORT_COLOR
+            flags |= SUPPORT_COLOR | SUPPORT_BRIGHTNESS
+        if self.device.supports_color_temperature or \
+                self.device.supports_tunable_white:
+            flags |= SUPPORT_COLOR_TEMP
         return flags
 
     async def async_turn_on(self, **kwargs):
         """Turn the light on."""
-        if ATTR_BRIGHTNESS in kwargs:
-            if self.device.supports_brightness:
-                await self.device.set_brightness(int(kwargs[ATTR_BRIGHTNESS]))
-        elif ATTR_HS_COLOR in kwargs:
-            if self.device.supports_color:
-                await self.device.set_color(color_util.color_hs_to_RGB(
-                    *kwargs[ATTR_HS_COLOR]))
+        brightness = kwargs.get(ATTR_BRIGHTNESS, self.brightness)
+        hs_color = kwargs.get(ATTR_HS_COLOR, self.hs_color)
+        mireds = kwargs.get(ATTR_COLOR_TEMP, self.color_temp)
+
+        update_brightness = ATTR_BRIGHTNESS in kwargs
+        update_color = ATTR_HS_COLOR in kwargs
+        update_color_temp = ATTR_COLOR_TEMP in kwargs
+
+        # always only go one path for turning on (avoid conflicting changes
+        # and weird effects)
+        if self.device.supports_brightness and \
+                (update_brightness and not update_color):
+            # if we don't need to update the color, try updating brightness
+            # directly if supported; don't do it if color also has to be
+            # changed, as RGB color implicitly sets the brightness as well
+            await self.device.set_brightness(brightness)
+        elif self.device.supports_color and \
+                (update_brightness or update_color):
+            # change RGB color (includes brightness)
+            # if brightness or hs_color was not yet set use the default value
+            # to calculate RGB from as a fallback
+            if brightness is None:
+                brightness = DEFAULT_BRIGHTNESS
+            if hs_color is None:
+                hs_color = DEFAULT_COLOR
+            await self.device.set_color(
+                color_util.color_hsv_to_RGB(*hs_color, brightness * 100 / 255))
+        elif self.device.supports_color_temperature and \
+                update_color_temp:
+            # change color temperature without ON telegram
+            kelvin = int(color_util.color_temperature_mired_to_kelvin(mireds))
+            if kelvin > self._max_kelvin:
+                kelvin = self._max_kelvin
+            elif kelvin < self._min_kelvin:
+                kelvin = self._min_kelvin
+            await self.device.set_color_temperature(kelvin)
+        elif self.device.supports_tunable_white and \
+                update_color_temp:
+            # calculate relative_ct from Kelvin to fit typical KNX devices
+            kelvin = int(color_util.color_temperature_mired_to_kelvin(mireds))
+            relative_ct = int(255 * (kelvin - self._min_kelvin) /
+                                    (self._max_kelvin - self._min_kelvin))
```

Comment:
pylint: Wrong continued indentation
https://travis-ci.org/home-assistant/home-assistant/jobs/490179399
```suggestion
                              (self._max_kelvin - self._min_kelvin))
```